api/src/context.rs

```diff
@@ -797,27 +806,168 @@ impl Context {
-        })
+            prioritized_gas_estimate: Some(self.next_bucket(min_gas_unit_price)),
+        }
+    }
+
+    fn cached_gas_estimation(&self, current_epoch: u64) -> Option<GasEstimation> {
+        let cache = self.gas_estimation_cache.read().unwrap();
+        if let Some(epoch) = cache.last_updated_epoch {
+            if let Some(time) = cache.last_updated_time {
+                if let Some(estimation) = cache.estimation {
+                    // TODO: config for updated time
+                    if epoch == current_epoch && time.elapsed().as_millis() < 500 {
+                        return Some(estimation);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub fn estimate_gas_price<E: InternalError>(
+        &self,
+        ledger_info: &LedgerInfo,
+    ) -> Result<GasEstimation, E> {
+        let max_block_history = 120;
+        let min_gas_unit_price = self.min_gas_unit_price(ledger_info)?;
+        let epoch = ledger_info.epoch.0;
+
+        // 0. Return cached result if it exists
+        if let Some(cached_gas_estimation) = self.cached_gas_estimation(epoch) {
+            return Ok(cached_gas_estimation);
+        }
+
+        let mut cache = self.gas_estimation_cache.write().unwrap();
+        // 1. Get the block metadata txns
+        let mut lookup_version = ledger_info.ledger_version.0;
+        let mut blocks = vec![];
+        // Skip the first block, which may be partial
+        if let Ok((first, _, block)) = self.db.get_block_info_by_version(lookup_version) {
+            if block.epoch() == epoch {
+                lookup_version = first.saturating_sub(1);
+            }
+        }
+        let mut cached_blocks_hit = false;
+        for _i in 0..max_block_history {
+            if cache
+                .min_inclusion_prices
+                .contains_key(&(epoch, lookup_version))
+            {
+                cached_blocks_hit = true;
+                break;
+            }
+            match self.db.get_block_info_by_version(lookup_version) {
+                Ok((first, last, block)) => {
+                    if block.epoch() != epoch {
+                        break;
+                    }
+                    lookup_version = first.saturating_sub(1);
+                    blocks.push((first, last));
+                    if lookup_version == 0 {
+                        break;
+                    }
+                },
+                Err(_) => {
+                    break;
+                },
+            }
+        }
+        if blocks.is_empty() && !cached_blocks_hit {
+            let estimation = self.default_gas_estimation(min_gas_unit_price);
+            cache.estimation = Some(estimation);
+            cache.last_updated_epoch = Some(epoch);
+            cache.last_updated_time = Some(Instant::now());
+            return Ok(estimation);
+        }
+        let remaining = max_block_history - blocks.len();
+
+        // 2. Get gas prices per block
```

Comment:
nit: should this code block be in a helper function?

Reply:
I guess it's a style issue - I'd rather not put this into a helper function as it has a few shared variables with the other parts and won't be reusable.